server/interceptor/interceptor.go

```diff
@@ -0,0 +1,170 @@
+	// Whitelisted to skip tree validation
+	switch req.(type) {
+	case *trillian.CreateTreeRequest:
+		return &rpcInfo{doesNotHaveTree: true}, nil
+	case *trillian.ListTreesRequest:
+		return &rpcInfo{doesNotHaveTree: true}, nil
+	}
+
+	var treeID int64
+	switch req := req.(type) {
+	case treeIDRequest:
+		treeID = req.GetTreeId()
+	case treeRequest:
+		treeID = req.GetTree().GetTreeId()
+	case logIDRequest:
+		treeID = req.GetLogId()
+	case mapIDRequest:
+		treeID = req.GetMapId()
+	default:
+		return nil, errors.Errorf(errors.Internal, "cannot retrieve treeID from request: %T", req)
+	}
+
+	serviceName, methodName, err := parseFullMethod(fullMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	var treeType trillian.TreeType
+	switch serviceName {
+	case "trillian.TrillianAdmin":
+		treeType = trillian.TreeType_UNKNOWN_TREE_TYPE // unrestricted
+	case "trillian.TrillianLog":
+		treeType = trillian.TreeType_LOG
+	case "trillian.TrillianMap":
+		treeType = trillian.TreeType_MAP
+	default:
+		return nil, errors.Errorf(errors.Internal, "cannot determine treeType for request: %T", req)
+	}
+
+	readonly := strings.HasPrefix(methodName, "Get") || strings.HasPrefix(methodName, "List")
+
+	return &rpcInfo{
+		treeID: treeID,
+		opts:   trees.GetOpts{TreeType: treeType, Readonly: readonly},
+	}, nil
+}
+
+// parseFullMethod returns the service and method names as separate strings, without a trailing
+// slash in either of them.
+func parseFullMethod(fullMethod string) (string, string, error) {
+	if !strings.HasPrefix(fullMethod, "/") {
+		return "", "", errors.Errorf(errors.Internal, "fullMethod must begin with '/': %v", fullMethod)
+	}
+	tmp := strings.Split(fullMethod[1:], "/")
+	if len(tmp) != 2 {
+		return "", "", errors.Errorf(errors.Internal, "unexpected number of components in fullMethod (%v != 2): %v", len(tmp), fullMethod)
+	}
+	return tmp[0], tmp[1], nil
+}
+
+type treeIDRequest interface {
+	GetTreeId() int64
+}
+
+type treeRequest interface {
+	GetTree() *trillian.Tree
+}
+
+type logIDRequest interface {
+	GetLogId() int64
+}
+
+type mapIDRequest interface {
+	GetMapId() int64
+}
+
+// Combine combines multiple unary interceptors. Interceptors are executed in the supplied order.
+// If an interceptor fails (non-nil error), processing will stop and the error will be returned.
+// If all interceptors succeed the handler will be called.
+// Contexts are propagated between calls: the context of the first interceptor, as passed in to the
+// handler function, is used for the second interceptor call and so on, until the handler is
+// invoked. This ensures that request-level variables, transmitted via contexts, behave properly.
+func Combine(interceptors ...grpc.UnaryServerInterceptor) grpc.UnaryServerInterceptor {
+	return func(initialCtx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		// Save and forward the ctx passed by interceptors, otherwise we lose the data they add.
```

Comment:
It took me a couple of attempts to follow what was going on here.  How about some comments along the lines of:
```go
// Build a handler function that just updates the scope's ctx variable.
ctxUpdater := func....

// Now run each of the interceptors in turn, using the ctxUpdater to accumulate
// changes to the context, and exit if any of the interceptors give an error.
for ...

// All the interceptors' error checks have passed, so finally run the real
// handler with the accumulated context.
```